pkg/kbld/builder/bazel/bazel.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package image
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os/exec"
+	"regexp"
+
+	ctlbdk "github.com/k14s/kbld/pkg/kbld/builder/docker"
+	"github.com/k14s/kbld/pkg/kbld/config"
+	ctllog "github.com/k14s/kbld/pkg/kbld/logger"
+)
+
+var (
+	// Example output that includes the final digest:
+	// INFO: Analyzed target //:simple-app (0 packages loaded, 0 targets configured).
+	// INFO: Found 1 target...
+	// Target //:simple-app up-to-date:
+	//   bazel-bin/simple-app-layer.tar
+	// INFO: Elapsed time: 0.148s, Critical Path: 0.00s
+	// INFO: 1 process: 1 internal.
+	// INFO: Build completed successfully, 1 total action
+	// INFO: Build completed successfully, 1 total action
+	// Loaded image ID: sha256:328b5f47550c85cea5284911ad4d284ce20e8240d61d2610eb6cb4aa8b43c19e
+	// Tagging 328b5f47550c85cea5284911ad4d284ce20e8240d61d2610eb6cb4aa8b43c19e as bazel:simple-app
+	bazelImageID = regexp.MustCompile("Loaded image ID: (sha256:)?([0-9a-z]+)")
+)
+
+type Bazel struct {
+	docker ctlbdk.Docker
+	logger ctllog.Logger
+}
+
+func NewBazel(docker ctlbdk.Docker, logger ctllog.Logger) Bazel {
+	return Bazel{docker: docker, logger: logger}
+}
+
+func (b *Bazel) Build(image, directory string, opts config.SourceBazelBuildOpts) (ctlbdk.DockerTmpRef, error) {
+
+	prefixedLogger := b.logger.NewPrefixedWriter(image + " | ")
+
+	prefixedLogger.Write([]byte(fmt.Sprintf("starting build (using bazel): %s\n", directory)))
+	defer prefixedLogger.Write([]byte("finished build (using bazel)\n"))
+
+	var imageID string
+	{
+		var stdoutBuf, stderrBuf bytes.Buffer
+
+		cmdArgs := []string{"run"}
+
+		if opts.Label == nil {
+			return ctlbdk.DockerTmpRef{}, fmt.Errorf("Expected label to be specified, but was not")
+		}
+
+		cmdArgs = append(cmdArgs, *opts.Label)
+
+		if opts.RawOptions != nil {
+			cmdArgs = append(cmdArgs, *opts.RawOptions...)
+		}
+
+		cmd := exec.Command("bazel", cmdArgs...)
+		cmd.Dir = directory
+		cmd.Stdout = io.MultiWriter(&stdoutBuf, prefixedLogger)
+		cmd.Stderr = io.MultiWriter(&stderrBuf, prefixedLogger)
+
+		err := cmd.Run()
+		if err != nil {
+			prefixedLogger.Write([]byte(fmt.Sprintf("error: %s\n", err)))
+			return ctlbdk.DockerTmpRef{}, err
+		}
+
+		matches := bazelImageID.FindStringSubmatch(stdoutBuf.String())
+		if len(matches) != 3 {
```

Comment:
im guessing it's possible that bazel output can have multiple images loaded? i think we should fail instead of just picking up first one.

Reply:
I'm not entirely sure of all the build outcomes for bazel, but in our use case there are multiple stages/layers in our bazel build where we do load an image which becomes the base image of the next stage/layer.

Something like:
```
path/to/project | Loaded image: bazel/image:image
```

However, the final stage output is always:
```
Loaded image ID: sha256:<digest>
```
which is what regex for here. The extra `ID` _seems_ unique enough for the final output.

Reply:
i was thinking if your bazel config contains something like this:

```
go_image(
    name = "simple-app-go-image",
    srcs = ["app.go"],
    importpath = "github.com/carvel-kbld/test/e2e/assets/simple-app",
)

go_image(
    name = "simple-app-go-image-2",
    srcs = ["app.go"],
    importpath = "github.com/carvel-kbld/test/e2e/assets/simple-app",
)
```

that are both leaf images.

Reply:
Note that:

`bazel build` builds _one or many_ targets:

```bash
$ bazel build :simple-app1 :simple-app2
INFO: Analyzed 2 targets (85 packages loaded, 7287 targets configured).
INFO: Found 2 targets...
INFO: Elapsed time: 4.869s, Critical Path: 3.13s
INFO: 68 processes: 24 internal, 44 darwin-sandbox.
INFO: Build completed successfully, 68 total actions
$ 
```

while `bazel run` builds and executes _exactly one_ target (and it's the first one mentioned):

```bash
$ bazel run :simple-app1 :simple-app2
INFO: Analyzed target //:simple-app1 (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target //:simple-app1 up-to-date:
  bazel-bin/simple-app1-layer.tar
INFO: Elapsed time: 0.171s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
INFO: Build completed successfully, 1 total action
8651333b21e7: Loading layer [==================================================>]  3.031MB/3.031MB
ba16d454860a: Loading layer [==================================================>]  15.44MB/15.44MB
ea465d14ddf7: Loading layer [==================================================>]  6.472MB/6.472MB
84ff92691f90: Loading layer [==================================================>]  10.24kB/10.24kB
Loaded image ID: sha256:9e5eeee36c61be12f84bfdff09a14805f33d7056c2dac8d6758d7bd04d25a424
Tagging 9e5eeee36c61be12f84bfdff09a14805f33d7056c2dac8d6758d7bd04d25a424 as bazel:simple-app1
$
```

Note that `bazel build` produces all of the artifacts (i.e. tarball, OCI image config, digests); it's that they are not loaded into the local image repo.

Will it be explicit/obvious to a user that the Bazel integration with `kbld` is performing a `bazel run` (rather than a `bazel build`)?

Reply:
>Will it be explicit/obvious to a user that the Bazel integration with kbld is performing a bazel run (rather than a bazel build)?

This was something I was worried about as well. `bazel run`was used here becase it was the only way to load the image into the local image repo, but `bazel run <target>` isn't limited to just image building (it can build and execute the built image if the last stage was something like `go_image(...)`). The final step that seemed safe was `container_image(...)` because it just loads the images, but does not execute them.

I'm not sure if this would be obvious to a user. Perhaps we can add a note here: https://carvel.dev/kbld/docs/latest/config/#sources about the underlying command being `bazel run` instead of `bazel build` and using `container_image(...)`.

Reply:
for docker source (https://carvel.dev/kbld/docs/latest/config/#docker) we have explicit build subsection. might make sense to intro this for bazel (e.g. bazel.run section)

Reply:
> while bazel run builds and executes exactly one target (and it's the first one mentioned):

the example i was providing with multiple container_images would still be considered a single target right?

Reply:
> > while bazel run builds and executes exactly one target (and it's the first one mentioned):
> 
> the example i was providing with multiple container_images would still be considered a single target right?

Yup, in this case, target can only be either `simple-app-go-image` or `simple-app-go-image-2`.


> for docker source (https://carvel.dev/kbld/docs/latest/config/#docker) we have explicit build subsection. might make sense to intro this for bazel (e.g. bazel.run section)
> 
> edit: oh looks like we already have build section under bazel, i think we should rename it to run.

I can change it to `run`. I'm assuming something like:
```
  bazel:
    # instead of "build:"
    run:
      target: :simple-app
```

Reply:
got it. yup on run.